src/tests/ll_verification.py

```diff
@@ -723,6 +723,269 @@ def ll_ddi_adv_bv_21_c(transport, upperTester, lowerTester, trace):
+
+        def collectChainedAdvData(packet):
+            advData = bytes([])
+            if packet.payload['SuperiorPackets'][0].type.name != 'ADV_EXT_IND':
+                advData = collectChainedAdvData(packet.payload['SuperiorPackets'][0])
+            advData += packet.payload['AD']
+            return advData
+
+        # Check AUX_CHAIN_INDs
+        # AdvMode set to 00b; The AUX_CHAIN_IND PDU shall include the ADI field matching the ADI field from AUX_ADV_IND
+        # and the AdvData field containing additional data. The AUX_CHAIN_IND PDU shall not include the AdvA,
+        # TargetA, TxPower, or SyncInfo fields
+        # The time between a PDU containing an AuxPtr field and the PDU to which it refers shall be greater than or equal to T_MAFS
+        for packet in packets.fetch(packet_filter=('AUX_CHAIN_IND')):
+            success = success and packet.payload['AdvMode'] == 0b00
+            success = success and 'AdvA' not in packet.payload
+            success = success and 'TargetA' not in packet.payload
+            success = success and 'TxPower' not in packet.payload
+            success = success and 'SyncInfo' not in packet.payload
+            success = success and 'ADI' in packet.payload
+            success = success and packet.payload['ADI'].SID == Sid
+            for superiorPacket in packet.payload['SuperiorPackets']:
+                success = success and packet.payload['ADI'].DID == superiorPacket.payload['ADI'].DID
+                success = success and packet.ts >= superiorPacket.ts + 300
+            if 'AuxPtr' not in packet.payload:
+                # Collect complete advertising data and check advertising data against input
+                completeAdvDataFound += 1
+                collectedAdvData = collectChainedAdvData(packet)
+                success = success and len(collectedAdvData) == round.DataLength
+                for i in range(round.DataLength):
+                    success = success and collectedAdvData[i] == advData[i]
+
+        if round.DataLength > 0:
+            # Check that we actually got the complete data (ie. that some AUX_CHAIN_INDs/AUX_ADV_INDs do not have an aux ptr)
+            success = success and completeAdvDataFound > 0
+
+        if round.MaxEvents != 0:
+            # Calculate number of advertising events by grouping ADV_EXT_INDs (a delay of more than 10 ms or 3 packets in the group means a new group)
+            numEvents = 0
+            groupSize = 0
+            lastTs = -10000
+            for packet in packets.fetch(packet_filter=('ADV_EXT_IND')):
+                if lastTs < packet.ts - 10000 or groupSize >= 3:
+                    groupSize = 1
+                    numEvents += 1
+                else:
+                    groupSize += 1
+                lastTs = packet.ts
+            # Verify that the IUT did not send more than Max_Extended_Advertising_Events advertising events
+            success = success and numEvents <= round.MaxEvents
```

Comment:
This seems like a heuristic that is tailored towards how the Zephyr controller currently schedules advertisements packets. How hard would it be to generalize this?

Reply:
It was an attempt to have a generalized counting; But I guess it can be improved by taking the channel number into account (a duplicate channel would also mean a new event). I'll add that